src/extension-runners/firefox-android.js

```diff
@@ -0,0 +1,653 @@
+      await adbClient.shell(selectedAdbDevice.id, [
+        'rm', '-rf', selectedArtifactsDir,
+      ]);
+    }
+
+    // Call all the registered clenaup callbacks.
+    for (const fn of this.cleanupCallbacks) {
+      try {
+        fn();
+      } catch (error) {
+        log.error(error);
+      }
+    }
+  }
+
+  // Private helper methods.
+
+  getDeviceProfileDir(): string {
+    return `${this.selectedArtifactsDir}/profile`;
+  }
+
+  printIgnoredParamsWarnings() {
+    if (this.params.profilePath) {
+      log.warn(
+        'Firefox for Android target does not support custom profile paths.'
+      );
+    }
+
+    if (this.params.keepProfileChanges) {
+      log.warn(
+        'Firefox for Android target does not support --keep-profile-changes.'
+      );
+    }
+
+    if (this.params.browserConsole) {
+      log.warn(
+        'Firefox for Android target does not support --browser-console option.'
+      );
+    }
+
+    if (this.params.preInstall) {
+      log.warn(
+        'Firefox for Android target does not support --pre-install option.'
+      );
+    }
+
+    if (this.params.startUrl) {
+      log.warn(
+        'Firefox for Android target does not support --start-url option.'
+      );
+    }
+  }
+
+  async adbDevicesDiscoveryAndSelect() {
+    const {adbClient} = this;
+    const {adbDevice} = this.params;
+    let devices = [];
+
+    try {
+      log.debug('Listing android devices');
+      devices = await adbClient.listDevices();
+    } catch (error) {
+      if (isErrorWithCode('ENOENT', error) &&
+          error.message.includes('spawn adb')) {
+        throw new UsageError(
+          'No adb executable has been found. ' +
+          'You can Use --adb-bin, --adb-host/--adb-port ' +
+          'to configure it manually if needed.');
+      }
+
+      throw error;
+    }
+
+    if (devices.length === 0) {
+      throw new UsageError('No Android device found through ADB.');
+    }
+
+    if (!adbDevice) {
+      const devicesMsg = devices.map((dev) => ` - ${dev.id}`).join('\n');
+      log.info(`\nAndroid devices found:\n${devicesMsg}`);
+      throw new UsageError(
+        'Select an android device using --android-device');
+    }
+
+    const foundDevices = devices.filter((device) => {
+      return device.id === adbDevice;
+    });
+
+    if (foundDevices.length === 0) {
+      const devicesMsg = JSON.stringify(devices);
+      throw new UsageError(
+        `Android Device not found: ${adbDevice} in ${devicesMsg}`);
+    }
+
+    this.selectedAdbDevice = foundDevices[0];
+    log.info(`Selected adb device: ${this.selectedAdbDevice.id}`);
+  }
+
+  async apkPackagesDiscoveryAndSelect() {
+    const {
+      adbClient,
+      selectedAdbDevice,
+      params: {
+        adb,
+        firefoxApk,
+      },
+    } = this;
+    // Discovery and select a Firefox for Android version.
+    const pmList = await adbClient.shell(selectedAdbDevice.id, [
+      'pm', 'list', 'packages',
+    ]).then(adb.util.readAll);
```

Comment:
@kumar303 I'm thinking that we should turn this `adbClient.shell(...).then(adb.util.readAll)` pattern into a reusable helper, I'm afraid that forgetting to drain the output from an `adbClient.shell` command can lead to bugs hard to debug (e.g. if the output of a previous `adbClient.shell` command can become part of the output read from another unrelated `adbClient.shell` command which reads the output using `adb.util.readAll`).